src/allmydata/client.py

```diff
@@ -179,42 +203,86 @@ def _sequencer(self):
         return seqnum, nonce
 
-    def init_introducer_clients(self):
-        self.introducer_furls = []
+    def load_connections_from_yaml(self, furl):
+        connections_filepath = FilePath(os.path.join(self.basedir, "private", "connections.yaml"))
+        if connections_filepath.exists():
+            exists = True
+            with connections_filepath.open() as f:
+                connections = yaml.load(f)
+                f.close()
+        else:
+            exists = False
+            connections = { 'introducers' : {},
+                            'servers' : {},
+                            'transport_plugins' : {
+                                'tcp' : {
+                                    'handler_module' : 'foolscap.connection_plugins',
+                                    'handler_name': 'DefaultTCP',
+                                    'parameters' : {}
+                                },
+                            },
+                            }
+            new_connections = connections.copy()
+            new_connections['introducers'][u'default'] = {}
+            new_connections['introducers']['default']['furl'] = furl
+            connections_filepath.setContent(yaml.dump(new_connections))
+        return connections, exists
+
+    def load_connections(self):
+        """
+        Load the connections.yaml file if it exists, otherwise
+        create a default configuration. Abort startup and report
+        an error to the user if the tahoe.cfg contains an introducer
+        FURL which is also found in the connections.yaml.
+        """
+        # read introducer from tahoe.cfg and abort + error an introducer furl is specified
+        # which is also found in our connections.yaml
+        self.introducer_furls = [] # XXX
+        tahoe_cfg_introducer_furl = self.get_config("client", "introducer.furl", None)
         self.warn_flag = False
-        # Try to load ""BASEDIR/private/introducers" cfg file
-        cfg = os.path.join(self.basedir, "private", "introducers")
-        if os.path.exists(cfg):
-            f = open(cfg, 'r')
-            for introducer_furl in f.read().split('\n'):
-                introducer_furl_stripped = introducer_furl.strip()
-                if introducer_furl_stripped.startswith('#') or not introducer_furl_stripped:
-                    continue
-                self.introducer_furls.append(introducer_furl_stripped)
-            f.close()
-        furl_count = len(self.introducer_furls)
-
-        # read furl from tahoe.cfg
-        ifurl = self.get_config("client", "introducer.furl", None)
-        if ifurl and ifurl not in self.introducer_furls:
-            self.introducer_furls.append(ifurl)
-            f = open(cfg, 'a')
-            f.write(ifurl)
-            f.write('\n')
-            f.close()
-            if furl_count > 1:
+
+        connections, connections_yaml_exists = self.load_connections_from_yaml(tahoe_cfg_introducer_furl)
```

Comment:
passing the (deprecated) `tahoe_cfg_introducer_furl` to `load_connections_from_yaml` seems weird. I think `load_connections_from_yaml` is trying to do 3 different things: check if the file exists, load it if it does, and otherwise populate the default connections dictionary using the deprecated furl. Perhaps these should be two or three functions? Also, we want to be able to populate the default introducer entry with a furl passed from the commandline in the future, not just from the deprecated `tahoe.cfg` option.